### ISSUE TITLE
Thread didn't execute?

### ISSUE THREAD

**Issue by ioquatix** [jruby/jruby#6142]
```
koyoko% jruby -v             
jruby 9.2.9.0 (2.5.7) 2019-10-30 458ad3e OpenJDK 64-Bit Server VM 25.242-b08 on 1.8.0_242-b08 +jit [linux-x86_64]
koyoko% uname -a
Linux koyoko.local 5.5.9-arch1-2 #1 SMP PREEMPT Thu, 12 Mar 2020 23:01:33 +0000 x86_64 GNU/Linux
```

In my live stream, talking about thread safety no less, I encountered some kind of issue in JRuby when using threads in code that should have been fine.

https://youtu.be/Wu9LRNOc5pQ?t=630

It appears that while we did construct a thread, it never actually executed any code.

Here is the code from the live stream above:

```ruby
#!/usr/bin/env ruby

def test(repeats = 1000)
	mutex = Mutex.new
	
	mutexes = {}
	threads = []
	
	threads = repeats.times.map do |i|
		Thread.new do
			mutex.synchronize do
				mutexes[i] ||= Mutex.new
			end
		end
	end
	
	threads.each(&:join)
	
	if threads.size != mutexes.size
		puts "There were #{threads.size} threads and #{mutexes.size} mutexes."
		
		repeats.times do |i|
			unless mutexes.include?(i)
				puts "Missing #{i}!"
			end
		end
	end
	
	missing = repeats.times.to_a - mutexes.keys
	
	unless missing.empty?
		puts "Missing: #{missing.inspect}"
	end
end

100.times{test}
```

I have been discussing the issue with @headius and he suggested it might be a problem with the native thread being GCed before starting:

```java
    public void start() {
        Thread thread = getThread();
        if (thread != null) thread.start();
        else throw new RuntimeException("WTF");
    }
```

and he was able to get it to fail by increasing GC frequency.

```
[] ~/projects/jruby $ jruby thread_mutexes.rb 
Unhandled Java exception: java.lang.RuntimeException: WTF
```

However I am not able to reproduce the issue, it seems very infrequent.

**Comment by headius** [comment]
Yeah my theory is just a theory, but it seems the most likely candidate right now.

In NativeThread, we hold a weak reference to the actual JVM Thread object. I believe this is getting collected at just the right moment in the video link provided (at 10:34) such that it silently fails to run the thread's body.

The failure, using the script in the video, happened toward the end of the run after spawning, joining, and walking away from tens of thousands of threads. It's certainly possible that @ioquatix "luckily" happened to have a GC run trigger between the creation of a RubyThread object and the starting of the native thread associated with it, such that the native thread object got collected before it could even start running.

The last snippit of code above contains a modification: raising an error if the weak reference to the thread is no longer valid. The original code contains no such error, allowing the start to silently succeed if the associated Thread object has already gone away.

I will commit a short term fix now that at least does a hard error, but the long term fix would be to make all Ruby threads be hard-referenced by their RubyThread object. Java threads we adopt should still be weak, so we don't keep their Thread objects (and related state) alive until JRuby cleans up its references (probably at exit).

**Comment by headius** [comment]
Simple fix is on master. Refactor can happen at our leisure.

**Comment by ioquatix** [comment]
Great thanks for your effort!

**Comment by rsim** [comment]
We are occasionally hit by this bug in our JRuby 9.2 application. I wanted to share a simpler test code which typically exposes this bug:
```ruby
require "concurrent-ruby"
def test_threads(repeats = 1000)
  counter = Concurrent::AtomicFixnum.new
  threads = repeats.times.map do
    thread = Thread.new { counter.increment }
    puts "Missing native thread" unless thread.to_java.getNativeThread
    thread
  end
  threads.each(&:join)
  if (missing = repeats - counter.value) > 0
    puts "Missing #{missing} threads"
  end
end
100.times { test_threads }
```

We plan to upgrade to JRuby 9.3 later but in the meantime we created a helper method which validates that the native thread is not garbage collected before using this thread:
```ruby
def new_thread(*args, &block)
  retries = 3
  while retries > 0
    thread = Thread.new(*args, &block)
    return thread if thread.to_java.getNativeThread
    retries -= 1
  end
  raise "failed to create a new thread"
end
```

**Comment by headius** [comment]
@rsim okay, thanks for the report. I will backport the fix to 9.2 if it has not made it there already and we will see about a new release at some point. We just did 9.2.20 however so it may be a little bit.
